sdk/src/wallet/mod.rs

```diff
@@ -481,52 +493,59 @@ impl<U: WalletIo> Wallet<U> {
     /// provided, will prompt for password from stdin.
     /// Stores the key in decrypted key cache and returns the alias of the key
     /// and a reference-counting pointer to the key.
-    pub fn derive_key_from_user_mnemonic_code(
+    pub fn derive_key_from_mnemonic_code(
         &mut self,
         scheme: SchemeType,
         alias: Option<String>,
         alias_force: bool,
-        derivation_path: Option<String>,
+        derivation_path: DerivationPath,
         mnemonic_passphrase: Option<(Mnemonic, Zeroizing<String>)>,
         password: Option<Zeroizing<String>>,
-    ) -> Result<Option<(String, common::SecretKey)>, GenRestoreKeyError> {
-        let parsed_derivation_path = derivation_path
-            .map(|p| {
-                let is_default = p.eq_ignore_ascii_case("DEFAULT");
-                if is_default {
-                    Ok(DerivationPath::default_for_scheme(scheme))
-                } else {
-                    DerivationPath::from_path_str(scheme, &p)
-                        .map_err(GenRestoreKeyError::DerivationPathError)
-                }
-            })
-            .transpose()?
-            .unwrap_or_else(|| DerivationPath::default_for_scheme(scheme));
-        if !parsed_derivation_path.is_compatible(scheme) {
-            println!(
-                "WARNING: the specified derivation path may be incompatible \
-                 with the chosen cryptography scheme."
-            )
-        }
-        println!("Using HD derivation path {}", parsed_derivation_path);
+    ) -> Result<(String, common::SecretKey), GenRestoreKeyError> {
         let (mnemonic, passphrase) =
             if let Some(mnemonic_passphrase) = mnemonic_passphrase {
                 mnemonic_passphrase
             } else {
                 (U::read_mnemonic_code()?, U::read_mnemonic_passphrase(false))
             };
         let seed = Seed::new(&mnemonic, &passphrase);
-
-        Ok(self.gen_and_store_key(
+        let sk = derive_hd_secret_key(
             scheme,
-            alias,
+            seed.as_bytes(),
+            derivation_path.clone(),
+        );
+
+        self.insert_keypair(
+            alias.unwrap_or_default(),
             alias_force,
-            Some((seed, parsed_derivation_path)),
+            sk.clone(),
             password,
-        ))
+            None,
+            Some(derivation_path),
+        )
+        .map(|alias| (alias, sk))
     }
 
-    /// Generate a new keypair and derive an implicit address from its public
+    /// Generate a spending key similarly to how it's done for keypairs
+    #[cfg(feature = "masp-tx-gen")]
```

Comment:
Why is this under a feature flag?

Reply:
You're right. This is a holdover from back when the WASM builds depended on the shared crate. Back then we were trying to prevent the WASM builds from indirectly depending on the `rand` and `rand_core` crates whilst not breaking MASP transaction generation that uses randomness.

But now, logically, if the SDK can generate secret keys without breaking MASP, then it should also be able to generate spending keys without breaking anything. Experimentally I've removed the flags and rebuilt the WASM binaries and their hashes haven't changed. I'm running the tests, but this removal is unlikely to have broken anything.